### ISSUE TITLE
Replace PyCharm code inspections by faster and easier tool?

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1157]
One problem with the PyCharm code inspections is that they are quite slow to run, and delay the CI tests.

Also, they might be difficult to run by others, although this has always been resolved in the end.

However, the requirements for an alternative are:

- It should actually be faster.
- It should be simple to run.
- It should cover mostly the same tests.

The last aspect is the tricky one. PyCharm goes well beyond PEP8 and checks for many standard Python conventions (many other PEPs). I don't know any other tool which even comes close.

(Related was the earlier discussion on `pycharm-inspect` here: #328)

**Comment by JackTemaki** [comment]
Maybe Flake8 with extensions that cover the non-PEP8 cases?

**Comment by JackTemaki** [comment]
https://github.com/DmytroLitvinov/awesome-flake8-extensions

**Comment by albertz** [comment]
Note that PyCharm also covers type checks. I think the most common other tool for that is mypy.

**Comment by albertz** [comment]
Note that PyCharm also covers certain error checks, e.g. that a function must exists, or an object attrib, object method, and many more things. This is a particular important feature because this already has saved us from otherwise faulty changes multiple times. (I.e. changes which would have breaked sth but where this break was not covered by some test.)

**Comment by albertz** [comment]
> > Note that PyCharm also covers type checks. I think the most common other tool for that is mypy.
> 
> At least for me I have type check warnings displayed for RETURNN that to do not fail the tests. I can not recommend mypy for RETURNN, as even in new projects this can cause a lot of overhead if you did not type hinted every single variable from the beginning. I needed quite the time to get used to mypy in another repository, which was definitely tiny compared to the size of RETURNN.

But I think it's good to have every single variable hinted. Just because it might mean lots of work does not mean we should not do it. Note that even much bigger projects like PyTorch successfully use MyPy.

Anyway, maybe MyPy is a separate discussion.

However, this issue here is about potentially replacing the PyCharm inspections. So if not MyPy, what are other alternatives which cover type checks in a similar way as PyCharm?

Actually, to be fair, PyCharm also has a couple of bugs regarding type checks, and we exclude some of the type checks in our `pycharm-inspect` tool.

**Comment by albertz** [comment]
Note that I tried both pylint and flake8 at some point and it was in a totally different league compared to PyCharm. Maybe the flake8 extensions you linked can somewhat catch up but I am not sure.

But I made this issue here also to be a bit more specific and explicitly collect cases we want to cover which are currently covered by PyCharm.

**Comment by albertz** [comment]
I now stumbled upon [Ruff - an extremely fast Python linter, written in Rust](https://github.com/charliermarsh/ruff). From a first glance, this looks interesting.

**Comment by NeoLegends** [comment]
Ruff is great, have been using it as a formatter in practice for a while now. It's speedy. There are some cases where it produces slightly different output compared to `black`.

**Comment by albertz** [comment]
Yes, Ruff is both a formatter and a linter. But this issue here is mostly about inspections/linting. So the question is, how does the Ruff linting compares to PyCharm inspections (basically my (incomplete) list in my original post). (The diff of the Ruff code formatter vs Black code formatting is not really so interesting to me. I would not care about that.)
